### PR TITLE
Fix calculated price for Jetpack Search in product description

### PR DESCRIPTION
**Pull request by CodeyGuyDylan**
## Summary

The current product description at `/wp-admin/admin.php?page=jetpack#/product/search` doesn't show the correct price. It doesn't account for sale coupons, and the search product's intro discount is done using a sale coupon. Additionally, we would like this page to recommend a JP Search tier based on how many records the site has.

#### Changes proposed in this Pull Request:

The product description component needs to consider sale coupons when calculated discounted price. Not only because the Search intro discount solely relies on it, but also because we want sales to be represented here when all products are on an additional sale for some holiday.

The product description specifically for JP search should take into account how many records the current site has, and recommend a plan based on that.

A data query component was created to query the `/jetpack/v4/site/products` endpoint. Previously, the products for these Product Descriptions were pulling their data from `/jetpack/v4/products` endpoint. The former seems to return correct pricing for the JP Search plans, taking into account post count on the site, where the latter did not. 

#### Jetpack product discussion
This came from [this discussion](https://github.com/Automattic/jetpack/pull/24461#issuecomment-1137768582)

#### Does this pull request change what data or activity we track or use?

My PR does not change any data tracking. I am using a different API endpoint to pull products into the Product Description, but this came from an existing Redux action, so I am not sure if this needs to be looked at by somebody or not

#### Testing instructions:

1. Start your local environment and start on branch `trunk`
2. Go to `/wp-admin/admin.php?page=jetpack#/product/search`, you will need to have at least Jetpack Free setup on your local environment and you cannot be in "offline mode." (I had a problem with being in offline mode when using localhost, so I used Jurassic Tube for this)
3. Confirm that the Search price on this page does not include the sale coupon intro discount, and that it does not change based on the number of records on your site
image
   - To test out the number of records, I imported records from [these files](https://github.com/WPTT/theme-test-data). This is only 50ish posts, so to get 100 I changed the titles and post-id's to all the posts via search and replace to make them unique again (if there is a better way to do this, please let me know 😄 
4. Switch to this branch: `fix/search-price-on-wp-product-page`
5. Refresh the page (make sure you have `jetpack watch plugins/jetpack` running to see the changes)
6. Confirm that the intro discount is now shown on this page and that the price changes depending on the number of posts on the site
image

Note: When importing posts, the site takes at least a few minutes to be re-indexed and recognize the new posts, so you may have to wait a bit to see the next tier of search pricing on this page after importing the content.

### CODE REVIEW

**Comment by CodeyGuyDylan** [comment]
I see the E2E tests timing out, not sure how normal that is or if something I did could have cause that, tried re-running it twice and it kept timing out. Somebody with more experience than me may know why that is happening or how to fix it 😄

**Comment by bindlegirl** [comment]
Works as advertised 👍

**Review by bindlegirl — APPROVED** [review]

File: projects/plugins/jetpack/_inc/client/components/data/query-site-products/index.jsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	static propTypes = {
+		isFetchingSiteProducts: PropTypes.bool,
+	};
+
+	static defaultProps = {
+		isFetchingSiteProducts: false,
+	};
+
+	componentDidMount() {
+		if ( ! this.props.isFetchingSiteProducts && isEmpty( this.props.siteProducts ) ) {
```
Comment (monsieur-z) [inline projects/plugins/jetpack/_inc/client/components/data/query-site-products/index.jsx:17]: The call to `isEmpty` means that we'll basically fetch site products only once per user session. I guess this is fine in most cases, but could potentially lead to some discrepancies between the plugin and server state. Is this a common pattern in the plugin codebase?

In Calypso, we've been trying to stop using `lodash` (I think it has been removed by now). I don't know if we're trying to do the same with the plugin.

File: projects/plugins/jetpack/_inc/client/product-descriptions/index.jsx [hunk 2]
```diff
@@ -44,7 +44,7 @@ const ProductDescriptions = props => {
 
 	return (
 		<>
-			<QueryProducts />
```
Comment (monsieur-z) [inline projects/plugins/jetpack/_inc/client/product-descriptions/index.jsx:47]: Do you know if there's any information queried in the products call that's not queried by the site products call?

**Review by monsieur-z — APPROVED** [review]
This tested well and code looks good. I left some questions/comments in the code, that you may or may not want to address.

E2E failing tests don't seem related to these changes.

File: projects/plugins/jetpack/_inc/client/components/data/query-site-products/index.jsx (hunk 1 above)
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/client/components/data/query-site-products/index.jsx:17]: @monsieur-z I copied essentially this exact pattern from the other data query components in the `components/data` directory. If we are trying to get rid of it in the plugin, I am thinking maybe stick with consistency for now and then go through all of these and edit them all at once at some point to remove the use of lodash?

File: projects/plugins/jetpack/_inc/client/product-descriptions/index.jsx (hunk 2 above)
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/client/product-descriptions/index.jsx:47]: Before this change, the product sent to the ProductDescription component was :
```json
{
	"title": "Jetpack Site Search",
	"slug": "jetpack_search",
	"key": "search",
	"description": "Help your site visitors find answers instantly so they keep reading and buying. Great for sites with a lot of content.",
	"features": [
		"Instant search and indexing",
		"Powerful filtering",
		"Supports 29 languages",
		"Spelling correction"
	],
	"available": true,
	"currencyCode": "USD",
	"showPromotion": true,
	"promotionPercentage": 50,
	"includedInPlans": [],
	"fullPrice": 59.95,
	"upgradeUrl": "https://jetpack.com/redirect/?source=jetpack-product-description-checkout&path=jetpack_search&site=munson1996.jurassic.tube"
}
```

After, it is:
```json
{
	"title": "Jetpack Site Search",
	"slug": "jetpack_search",
	"key": "search",
	"description": "Help your site visitors find answers instantly so they keep reading and buying. Great for sites with a lot of content.",
	"features": [
		"Instant search and indexing",
		"Powerful filtering",
		"Supports 29 languages",
		"Spelling correction"
	],
	"available": true,
	"currencyCode": "USD",
	"showPromotion": true,
	"promotionPercentage": 50,
	"includedInPlans": [],
	"fullPrice": 59.95,
	"saleCoupon": {
		"start_date": "2022-06-20 00:00:00",
		"expires": "2023-06-20 00:00:00",
		"discount": 50,
		"purchase_types": [
			3
		],
		"product_ids": [
			2104
		],
		"allowed_for_domain_transfers": false,
		"allowed_for_renewals": false,
		"allowed_for_new_purchases": true,
		"code": "b07ac1e15cb15678"
	},
	"upgradeUrl": "https://jetpack.com/redirect/?source=jetpack-product-description-checkout&path=jetpack_search&site=munson1996.jurassic.tube"
}
```

It appears to have all of the values as before plus a few extra 😄

**Comment by CodeyGuyDylan** [comment]
> This tested well and code looks good. I left some questions/comments in the code, that you may or may not want to address.
> 
> E2E failing tests don't seem related to these changes.

After reading through your comments, I believe we are good to merge 😄

**Pull request merged**